File: lib/cfr/builtins/libs/dart/ui/fontFeature.dart
import 'dart:core';
import 'dart:ui';

import 'package:hydro_sdk/cfr/runtimeSupport.dart';

class VMManagedFontFeature extends VMManagedBox<FontFeature> {
  VMManagedFontFeature(
      {required this.table, required this.vmObject, required this.hydroState})
      : super(
          table: table,
          vmObject: vmObject,
          hydroState: hydroState,
        ) {
    table['feature'] = vmObject.feature;
    table['value'] = vmObject.value;
    table['getHashCode'] =
        makeLuaDartFunc(func: (List<dynamic> luaCallerArguments) {
      return [
        vmObject.hashCode,
      ];
    });
    table['toString'] =
        makeLuaDartFunc(func: (List<dynamic> luaCallerArguments) {
      return [
        vmObject.toString(),
      ];
    });
  }

  final HydroTable table;

  final HydroState hydroState;

  final FontFeature vmObject;
}

class RTManagedFontFeature extends FontFeature implements Box<FontFeature> {
  RTManagedFontFeature(String feature$, int value$,
      {required this.table, required this.hydroState})
      : super(
          feature$,
          value$,
        ) {
    table['vmObject'] = vmObject;
    table['unwrap'] = makeLuaDartFunc(func: (List<dynamic> luaCallerArguments) {
      return [unwrap()];
    });
    table['feature'] = feature;
    table['value'] = value;
    table['_dart_getHashCode'] =
        makeLuaDartFunc(func: (List<dynamic> luaCallerArguments) {
      return [super.hashCode];
    });
    table['_dart_toString'] =
        makeLuaDartFunc(func: (List<dynamic> luaCallerArguments) {
      return [super.toString()];
    });
  }

  final HydroTable table;

  final HydroState hydroState;

  FontFeature unwrap() => this;
  FontFeature get vmObject => this;
  @override
  int get hashCode {
    Closure closure = table["getHashCode"];
    return closure.dispatch([table], parentState: hydroState)[0];
  }

  @override
  String toString() {
    Closure closure = table["__tostring"];
    return closure.dispatch([table], parentState: hydroState)[0];
  }
}

void loadFontFeature(
    {required HydroState hydroState, required HydroTable table}) {
  table['fontFeature'] =
      makeLuaDartFunc(func: (List<dynamic> luaCallerArguments) {
    return [
      RTManagedFontFeature(luaCallerArguments[1], luaCallerArguments[2],
          table: luaCallerArguments[0], hydroState: hydroState)
    ];
  });
  table['fontFeatureEnable'] =
      makeLuaDartFunc(func: (List<dynamic> luaCallerArguments) {
    return [
      maybeBoxObject<FontFeature>(
          object: FontFeature.enable(luaCallerArguments[1]),
          hydroState: hydroState,
          table: HydroTable()),
    ];
  });
  table['fontFeatureDisable'] =
      makeLuaDartFunc(func: (List<dynamic> luaCallerArguments) {
    return [
      maybeBoxObject<FontFeature>(
          object: FontFeature.disable(luaCallerArguments[1]),
          hydroState: hydroState,
          table: HydroTable()),
    ];
  });
  table['fontFeatureRandomize'] =
      makeLuaDartFunc(func: (List<dynamic> luaCallerArguments) {
    return [
      maybeBoxObject<FontFeature>(
          object: FontFeature.randomize(),
          hydroState: hydroState,
          table: HydroTable()),
    ];
  });
  table['fontFeatureStylisticSet'] =
      makeLuaDartFunc(func: (List<dynamic> luaCallerArguments) {
    return [
      maybeBoxObject<FontFeature>(
          object: FontFeature.stylisticSet(luaCallerArguments[1]),
          hydroState: hydroState,
          table: HydroTable()),
    ];
  });
  table['fontFeatureSlashedZero'] =
      makeLuaDartFunc(func: (List<dynamic> luaCallerArguments) {
    return [
      maybeBoxObject<FontFeature>(
          object: FontFeature.slashedZero(),
          hydroState: hydroState,
          table: HydroTable()),
    ];
  });
  table['fontFeatureOldstyleFigures'] =
      makeLuaDartFunc(func: (List<dynamic> luaCallerArguments) {
    return [
      maybeBoxObject<FontFeature>(
          object: FontFeature.oldstyleFigures(),
          hydroState: hydroState,
          table: HydroTable()),
    ];
  });
  table['fontFeatureProportionalFigures'] =
      makeLuaDartFunc(func: (List<dynamic> luaCallerArguments) {
    return [
      maybeBoxObject<FontFeature>(
          object: FontFeature.proportionalFigures(),
          hydroState: hydroState,
          table: HydroTable()),
    ];
  });
  table['fontFeatureTabularFigures'] =
      makeLuaDartFunc(func: (List<dynamic> luaCallerArguments) {
    return [
      maybeBoxObject<FontFeature>(
          object: FontFeature.tabularFigures(),
          hydroState: hydroState,
          table: HydroTable()),
    ];
  });
  registerBoxer<FontFeature>(boxer: (
      {required FontFeature vmObject,
      required HydroState hydroState,
      required HydroTable table}) {
    return VMManagedFontFeature(
        vmObject: vmObject, hydroState: hydroState, table: table);
  });
}

File: runtime/dart/ui/fontFeature.ts
declare const dart: {
    ui: {
        fontFeature: (
            this: void,
            fontFeature: IFontFeature,
            feature: string,
            value: number
        ) => IFontFeature;
        fontFeatureEnable: (feature: string) => IFontFeature;
        fontFeatureDisable: (feature: string) => IFontFeature;
        fontFeatureRandomize: () => IFontFeature;
        fontFeatureStylisticSet: (value: number) => IFontFeature;
        fontFeatureSlashedZero: () => IFontFeature;
        fontFeatureOldstyleFigures: () => IFontFeature;
        fontFeatureProportionalFigures: () => IFontFeature;
        fontFeatureTabularFigures: () => IFontFeature;
    };
};
export interface IFontFeature {
    feature: string;
    value: number;
    getHashCode: () => number;
    toString: () => string;
}
export class FontFeature {
    public static _kEncodedSize = 8;
    public readonly feature: string = undefined as any;
    public readonly value: number = undefined as any;
    public constructor(feature: string, value: number = 1) {
        dart.ui.fontFeature(this, feature, value);
    }
    public static enable(feature: string): IFontFeature {
        return dart.ui.fontFeatureEnable(feature);
    }
    public static disable(feature: string): IFontFeature {
        return dart.ui.fontFeatureDisable(feature);
    }
    public static randomize(): IFontFeature {
        return dart.ui.fontFeatureRandomize();
    }
    public static stylisticSet(value: number): IFontFeature {
        return dart.ui.fontFeatureStylisticSet(value);
    }
    public static slashedZero(): IFontFeature {
        return dart.ui.fontFeatureSlashedZero();
    }
    public static oldstyleFigures(): IFontFeature {
        return dart.ui.fontFeatureOldstyleFigures();
    }
    public static proportionalFigures(): IFontFeature {
        return dart.ui.fontFeatureProportionalFigures();
    }
    public static tabularFigures(): IFontFeature {
        return dart.ui.fontFeatureTabularFigures();
    }
    private readonly _dart_getHashCode: () => number = undefined as any;
    private readonly _dart_toString: () => string = undefined as any;
    public getHashCode(): number {
        return this._dart_getHashCode();
    }
    public toString(): string {
        return this._dart_toString();
    }
}
